functions/ocr/ocr-translate-text/src/main/java/functions/OcrTranslateText.java

```diff
@@ -97,5 +98,14 @@ public void accept(PubSubMessage pubSubMessage, Context context) {
       LOGGER.log(Level.SEVERE, "Error publishing translation save request: " + e.getMessage(), e);
     }
   }
+
+  private static String getenv(String name) {
```

Comment:
IMO we should still throw an error here - failing the deployment seems OK to me, as long as the user can find ouy why.

Reply:
I agree in principle, but in practice the current experience is pretty bad. The deployment fails with a cryptic message `Error setting up the execution environment for your function`. If that gets changed so it says something like `Deployment succeeded but a test execution failed`, or if we modify the API so that a background function can return an HTTP status other than 500, then we could reconsider.

Reply:
So - are you saying we should throw an Exception at invocation-time instead?

(IIUC we **have** to throw the exception somewhere - it doesn't make sense to use "default values" here, because [IMO] there _are_ no sensible "default values" for things like Pub/Sub topic names.)

Reply:
The problem is that if you deploy this sample without defining the required environment variables, currently the deployment will fail with the cryptic message I mentioned. That's because of the "health check" that gets run as the last step of deployment. I'm suggesting that a logged message is a better experience. If we were able to fail with an HTTP 400 that would be better still, since it wouldn't cause the health check to fail but would still show the problem clearly. Unfortunately that's not possible with the current API.

Reply:
I believe we _can_ log the error message _and then_ throw the exception though, right? (Or will that "swallow" the error message?)

Reply:
If we log a message and throw an exception, the message should indeed get logged, but the exception will result in a 500 status which will cause the health check to fail and the deployment to end with `Error setting up the execution environment for your function`. Users are not necessarily going to think to look in the logs at that point.

In fact, the current situation is worse than that because the null values will cause a NullPointerException in the static initializer, which means that the function class can't be loaded, and we'll give you a really obscure error message in the logs. But even fixing that NPE still leads to a poor user experience if you don't have the env vars.

Reply:
> If we log a message and throw an exception, the message should indeed get logged, but the exception will result in a 500 status which will cause the health check to fail and the deployment to end with Error setting up the execution environment for your function. **Users are not necessarily going to think to look in the logs at that point.**

Fair point - if @lesv is cool with this approach, we can merge as-is. We should still file an issue to revise this sample once we have proper error logging, though.

Reply:
You can add info about this in the README and the docs.  They will want to know these things for their own apps.

Reply:
A "What might go wrong" section is always helpful.  Explaining common errors is always a good thing.  If eng complains, you can explain how common the error is and ask them to fix it.

Reply:
Les - this is likely a temporary problem, so it's probably not worth adding specific information to the docs.

(There's probably an internal bug tracking this somewhere, but I can't seem to find it. Will post it here or ping you when I do.)